### ISSUE TITLE
database.compact() failed assertion

### ISSUE THREAD

**Issue by gmart7t2** [cberner/redb#657]
I have a 227 GB redb file and tried compacting it:

```
thread 'main' panicked at 'assertion failed: `(left != right)`
  left: `996063`,
 right: `996063`', /home/admin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/redb-1.0.5/src/tree_store/page_store/base.rs:76:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

real	6m21.195s
user	0m14.096s
sys	0m50.920s
```

I'm using redb version 1.0.5.

**Comment by gmart7t2** [comment]
Oh, they're not meant to be equal... Now I see

**Comment by gmart7t2** [comment]
I ran it again to get a backtrace:

```
thread 'main' panicked at 'assertion failed: `(left != right)`
  left: `996063`,
 right: `996063`', /home/admin/.cargo/registry/src/index.crates.io-6f17d22bba15001f/redb-1.0.5/src/tree_store/page_store/base.rs:76:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: redb::tree_store::btree::UntypedBtreeMut::relocate_helper
   5: redb::tree_store::btree::UntypedBtreeMut::relocate_helper
   6: redb::tree_store::btree::UntypedBtreeMut::relocate_helper
   7: redb::tree_store::btree::UntypedBtreeMut::relocate_helper
   8: redb::tree_store::btree::UntypedBtreeMut::relocate_helper
   9: redb::tree_store::table_tree::TableTree::compact_tables
  10: redb::transactions::WriteTransaction::compact_pages
  11: redb::db::Database::compact
  12: ord::subcommand::compact::run
  13: ord::subcommand::Subcommand::run
  14: ord::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

real	6m8.841s
user	0m14.704s
sys	0m49.394s
```

**Comment by cberner** [comment]
Uh oh. Do you have any way of reproducing it?

**Comment by cberner** [comment]
I've been able to reproduce it with the fuzzer. Not sure exactly what's wrong yet, but I'll take a look

**Comment by cberner** [comment]
Oh, are you using persistent savepoints? I think there's a bug where compact() will panic if there are any persistent savepoints in the db :/

**Comment by gmart7t2** [comment]
> Oh, are you using persistent savepoints? I think there's a bug where compact() will panic if there are any persistent savepoints in the db :/

I'm not using any savepoints at all. 

I have been trying to make a smaller database that demonstrates the issue. I deleted all but one of the tables, and deleted 90% of the 66 million rows from the remaining table, and it still fails when I try to compact it. 

The problem is that none of this deletion has decreased the database file at all, and when I gzip the database file it is actually bigger now than when I gzipped it before I started deleting things from it.

I guess since you have been able to reproduce the issue I don't need to keep trying to.
